Document HTTP client enhancements (#3483)
tarantool/doc: doc/reference/reference_lua/http.rst

Cookies
~~~~~~~

You can add cookies to the request using the :ref:`headers <request_options-headers>` option:

..  literalinclude:: /code_snippets/test/http_client/get_add_cookie_test.lua
    :language: lua
    :lines: 1-7

To learn how to obtain cookies passed in the ``Set-Cookie`` response header, see :ref:`Response cookies <response_cookies>`.

.. _request_body:

Body
~~~~

.. _request_serialization:

Serialization
*************

The HTTP client automatically serializes the content in a specific format when sending a request based on the specified ``Content-Type`` header.
By default, the client uses the ``application/json`` content type and sends data serialized as JSON:

..  literalinclude:: /code_snippets/test/http_client/post_json_test.lua
    :language: lua
    :lines: 1-6

The body for the request above might look like this:

.. code-block:: console

    {
        "user_id": 123,
        "user_name": "John Smith"
    }

To send data in a YAML or MsgPack format, set the ``Content-Type`` header explicitly to ``application/yaml`` or ``application/msgpack``, for example:

..  literalinclude:: /code_snippets/test/http_client/post_yaml_test.lua
    :language: lua

>     :lines: 1-10

Reviewer: please add an output here too, same as for example with default content-type

..  literalinclude:: /code_snippets/test/http_client/get_cookies_test.lua
    :language: lua
    :lines: 1-3

.. _response_body:

Response body
~~~~~~~~~~~~~

.. _response_deserialization:

Deserialization
***************

The HTTP client can deserialize response data to a Lua object based on the ``Content-Type`` response header value.
To deserialize data, call the :ref:`response_object.decode() <response_object-decode>` method.
In the example below, the JSON response is deserialized into a Lua object:

..  literalinclude:: /code_snippets/test/http_client/get_json_test.lua
    :language: lua
    :lines: 1-4

The following content types are supported out of the box:

* ``application/json``
* ``application/msgpack``
* ``application/yaml``

If the response doesn't have the ``Content-Type`` header, the client uses ``application/json``.

To deserialize other content types, you need to provide a custom deserializer
using the :ref:`client_object.decoders <client_object-decoders>` property.
In the example below, ``application/xml`` responses are decoded using the
``luarapidxml`` library:

..  literalinclude:: /code_snippets/test/http_client/get_xml_test.lua
    :language: lua

>     :lines: 1-12

Reviewer: an output for snippet would be useful here

        In the worst case, libcurl creates a new socket for every request,
        even if all requests go to the same host.

    ..  _client_options-max_total_connections:

    .. data:: max_total_connections

        Specifies the maximum number of active connections.
        This option affects libcurl `CURLMOPT_MAX_TOTAL_CONNECTIONS <https://curl.haxx.se/libcurl/c/CURLMOPT_MAX_TOTAL_CONNECTIONS.html>`_.

        **Note**

        You may want to control the maximum number of sockets that a particular HTTP client uses simultaneously.
        If a system passes many requests to distinct hosts, then libcurl cannot reuse sockets.
        In this case, setting ``max_total_connections`` may be useful
        since it causes curl to avoid creating too many sockets, which would not be used anyway.

..  _client_object:

client_object
~~~~~~~~~~~~~

..  class:: client_object

    An HTTP client instance that exposes the API for :ref:`making requests <making_requests>`.
    To create the client, call :ref:`http.client.new() <http-new>`.

    ..  _client_object-request:

    ..  method:: request(method, url, body, opts)

        Make an HTTP request and receive a response.

>         :param string method: a request HTTP method (``GET``, ``POST``, ``PUT``, ``PATCH``, ``OPTIONS``, ``HEAD``, ``DELETE``, ``TRACE``, ``CONNECT``)

Reviewer 2: ```suggestion
        :param string method: a request HTTP method. Possible values: ``GET``, ``POST``, ``PUT``, ``PATCH``, ``OPTIONS``, ``HEAD``, ``DELETE``, ``TRACE``, ``CONNECT``.
```

        :param string url: a request URL, for example, ``https://httpbin.org/get``
        :param string body: a request body (see :ref:`Body <request_body>`)
        :param table opts: request options (see :ref:`request_options <request_options>`)

        :return:

>             This method might return one of the following objects:

Reviewer 2: ```suggestion
            This method returns one of the following objects:
```

    ..  _request_options-proxy:

    .. data:: proxy

        A proxy server host or IP address, or ``''``.

        *   If ``proxy`` is a host or IP address, then it may begin with a scheme,
            for example, ``https://`` for an HTTPS proxy or ``http://`` for an HTTP proxy.
        *   If ``proxy`` is set to ``''`` an empty string, then proxy use is disabled,
            and no proxy-related environment variable is used.
        *   If ``proxy`` is not set, then a proxy-related environment variable may be used, such as
            ``HTTP_PROXY`` or ``HTTPS_PROXY`` or ``FTP_PROXY``, or ``ALL_PROXY`` if the
            protocol can be any protocol.

        :rtype: string

        **See also:** `CURLOPT_PROXY <https://curl.haxx.se/libcurl/c/CURLOPT_PROXY.html>`_

    ..  _request_options-proxy_port:

    .. data:: proxy_port

        A proxy server port.
        The default is ``443`` for an HTTPS proxy and ``1080`` for a non-HTTPS proxy.

        :rtype: integer

        **See also:** `CURLOPT_PROXYPORT <https://curl.haxx.se/libcurl/c/CURLOPT_PROXYPORT.html>`_

    ..  _request_options-proxy_user_pwd:

    .. data:: proxy_user_pwd

>         A proxy server username and/or password.

Reviewer 2: ```suggestion
        A proxy server username and password.
```

    .. data:: proto

        An HTTP protocol version.

        :rtype: string

    ..  _response_object-decode:

    .. function:: decode()

        Decode the response body to a Lua object based on the content type.

        :return: a decoded body
        :rtype: table

        **See also:** :ref:`Deserialization <response_deserialization>`

        **Since:** :doc:`2.11.0 </release/2.11.0>`

.. _io_object:

io_object
~~~~~~~~~

..  class:: io_object

    An IO object used to read or write data in chunks.
    To get an IO object instead of the full response (:ref:`response_object <response_object>`), you need to set the :ref:`chunked <request_options-chunked>` request option to ``true``.

    **Since:** :doc:`2.11.0 </release/2.11.0>`

    .. _io_object-read:

    .. function:: read(chunk[, timeout])
                  read(delimiter[, timeout])
                  read({chunk = chunk, delimiter = delimiter}[, timeout])

        Read request data in chunks of a specified length or up to a specific delimiter.

        :param integer chunk: the maximum number of bytes to read
        :param string delimiter: the delimiter used to stop reading data

>         :param integer timeout: the number of seconds to wait (the default is ``10``)

Reviewer 2: ```suggestion
        :param integer timeout: the number of seconds to wait. The default is ``10``.
```

        :return: A chunk of read data. Returns an empty string if there is nothing more to read.
        :rtype: string

        **See also:** :ref:`Streaming download <response_streaming_download>`

    .. _io_object-write:

    .. function:: write(data[, timeout])

        Write the specified chunk of data.

        :param table data: data to be written

>         :param integer timeout: the number of seconds to wait (the default is ``10``)

Reviewer 2: ```suggestion
        :param integer timeout: the number of seconds to wait. The default is ``10``.
```

        **See also:** :ref:`Streaming upload <request_streaming_upload>`

    .. _io_object-finish:

    .. function:: finish([timeout])

>         Finish reading/writing data.

Reviewer 2: ```suggestion
        Finish reading or writing data.
```

>         :param integer timeout: the number of seconds to wait (the default is ``10``)

Reviewer 2: ```suggestion
        :param integer timeout: the number of seconds to wait. The default is ``10``.
```

..  _http-module:

Module http
===========

..  module:: http.client

The ``http`` module, specifically the ``http.client`` submodule,
provides the functionality of an HTTP client with support for HTTPS and keepalive.
The HTTP client uses the `libcurl <https://curl.haxx.se/libcurl/>`_ library under the hood and
takes into account the `environment variables <https://curl.haxx.se/libcurl/c/libcurl-env.html>`_ libcurl understands.

.. _creating_client:

Creating a client
-----------------

> To create the HTTP client, call the :ref:`http.client.new() <http-new>` function:

Reviewer 3: ```suggestion
To create an HTTP client, call the :ref:`http.client.new() <http-new>` function:
```

..  literalinclude:: /code_snippets/test/http_client/get_test.lua
    :language: lua
    :lines: 1

Optionally, this function can accept specific client configuration options.

.. _making_requests:

Making requests
---------------

After :ref:`creating the client <creating_client>`, you can make HTTP requests.

.. _request_http_method:

HTTP method
~~~~~~~~~~~

> The main way of making HTTP requests is the :ref:`request <client_object-request>` method that can take a URL as a parameter.

Reviewer 3: _can take_ or always takes?
Also, _which_ would be better here than _that_ because this is an unrestrictive caluse.

Author: always takes, thanks. Replaced with one sentence:
```
The main way of making HTTP requests is the :ref:`request <client_object-request>` method, which accepts the following arguments:
```

This method accepts the following arguments:

*   An HTTP method, such as ``GET``, ``POST``, ``PUT``, and so on.
*   A request URL.
*   (Optional) A request body for the ``POST``, ``PUT``, and ``PATCH`` methods.

> *   (Optional) Request options, such as request headers, SSL settings, and so on.

Reviewer 2: ```suggestion
*   an HTTP method, such as ``GET``, ``POST``, ``PUT``, and so on
*   a request URL
*   (optional) a request body for the ``POST``, ``PUT``, and ``PATCH`` methods
*   (optional) request options, such as request headers, SSL settings, and so on
```

Headers
~~~~~~~

To add headers to the request, use the :ref:`headers <request_options-headers>` option:

..  literalinclude:: /code_snippets/test/http_client/get_add_header_test.lua
    :language: lua
    :lines: 1-8

.. _request_cookies:

Cookies
~~~~~~~

You can add cookies to the request using the :ref:`headers <request_options-headers>` option:

..  literalinclude:: /code_snippets/test/http_client/get_add_cookie_test.lua
    :language: lua
    :lines: 1-7

To learn how to obtain cookies passed in the ``Set-Cookie`` response header, see :ref:`Response cookies <response_cookies>`.

.. _request_body:

Body
~~~~

.. _request_serialization:

Serialization
*************

The HTTP client automatically serializes the content in a specific format when sending a request based on the specified ``Content-Type`` header.
By default, the client uses the ``application/json`` content type and sends data serialized as JSON:

..  literalinclude:: /code_snippets/test/http_client/post_json_test.lua
    :language: lua
    :lines: 1-6

The body for the request above might look like this:

.. code-block:: console

    {
        "user_id": 123,
        "user_name": "John Smith"
    }

> To send data in a YAML or MsgPack format, set the ``Content-Type`` header explicitly to ``application/yaml`` or ``application/msgpack``, for example:

Reviewer 3: ```suggestion
To send data in the YAML or MsgPack format, set the ``Content-Type`` header explicitly to ``application/yaml`` or ``application/msgpack``, for example:
```

..  literalinclude:: /code_snippets/test/http_client/post_yaml_test.lua
    :language: lua
    :lines: 1-10

In this case, the request body is serialized to YAML:

.. code-block:: console

    user_id: 123
    user_name: John Smith

.. _request_form_parameters:

Form parameters
***************

To send form parameters using the ``application/x-www-form-urlencoded`` type,
use the :ref:`params <request_options-params>` option:

..  literalinclude:: /code_snippets/test/http_client/post_form_params_test.lua
    :language: lua
    :lines: 1-5

.. _request_streaming_upload:

Streaming upload

> ****************

Reviewer 4: I would like to highlight here that a client can optionally set the `Content-Length` header manually:

https://github.com/tarantool/tarantool/blob/53c94bc79ecfc0451bc3ffa57ffb88e3217671ff/test/app-luatest/http_client_test.lua#L1225

In that case a regular HTTP POST (but the client code will continue to send data in chunks) method will be used instead of  https://en.wikipedia.org/wiki/Chunked_transfer_encoding .

Author: Setting the `Content-Length` header still uses chunked encoding when making a request, here is the issue:
https://github.com/tarantool/tarantool/issues/8744

Won't add this info to docs for now.

The HTTP client supports chunked writing of request data.

> This can be achieved as follows:

Reviewer 3: Nit: sounds a bit too formal, maybe a simpler wording?

1.  Set the :ref:`chunked <request_options-chunked>` option to ``true``.
    In this case, a request method returns :ref:`io_object <io_object>` instead of :ref:`response_object <response_object>`.
2.  Use the :ref:`io_object.write() <io_object-write>` method to write a chunk of data.
3.  Call the :ref:`io_object.finish() <io_object-finish>` method to finish writing data and make a request.

The example below shows how to upload data in two chunks:

..  literalinclude:: /code_snippets/test/http_client/post_stream_test.lua
    :language: lua
    :lines: 1-10

.. _receiving_responses:

Receiving responses
-------------------

> All methods used to :ref:`make an HTTP request <making_requests>` (``request``, ``get``, ``post``, etc.) receive :ref:`response_object <response_object>`.

Reviewer 3: ```suggestion
All methods that are used to :ref:`make an HTTP request <making_requests>` (``request``, ``get``, ``post``, etc.) receive :ref:`response_object <response_object>`.
```

Nit: just "used to" may be confused with the "<noun> used to <verb>" form by non-native speakers.

> ``response_object`` exposes the API required to get a response body and obtain response parameters, such as a status code, headers, and so on.

Reviewer 3: ```suggestion
``response_object`` exposes the API required to get a response body and obtain response parameters, such as its status code, headers, and so on.
```

.. _response_status_code:

Status code
~~~~~~~~~~~

> To a response's status code and text, use the :ref:`response_object.status <response_object-status>` and :ref:`response_object.reason <response_object-reason>` options, respectively:

Reviewer 3: missing verb?

Author: yep, thanks!

            -   An IO object used to read/write data in chunks

..  _http-new:

http.client.new()
~~~~~~~~~~~~~~~~~

..  function:: new([options])

    Create an HTTP client instance.

    :param table options: configuration options of the client (see :ref:`client_options <client_options>`)

    :return: a new HTTP client instance (see :ref:`client_object <client_object>`)
    :rtype:  userdata

    **Example**

    ..  literalinclude:: /code_snippets/test/http_client/get_test.lua
        :language: lua
        :lines: 1

..  _client_options:

client_options
~~~~~~~~~~~~~~

..  class:: client_options

    Configuration options of the client.
    These options can be passed to the :ref:`http.client.new() <http-new>` function.

    ..  _client_options-max_connections:

    .. data:: max_connections

        Specifies the maximum number of entries in the cache.
        This option affects libcurl `CURLMOPT_MAXCONNECTS <https://curl.haxx.se/libcurl/c/CURLMOPT_MAXCONNECTS.html>`_.
        The default is ``-1``.

        **Example**

        ..  code-block:: lua

            local http_client = require('http.client').new({max_connections = 5})

>         **Note**

Reviewer 3: why not an actual rst note?
`.. note::`

Author: I think the actual note for such a long description may distract the reader here. I agree that it is better to use the feature designed for such a use case, but two huge notes look worse here, I tried )

        In the worst case, libcurl creates a new socket for every request,
        even if all requests go to the same host.

    ..  _client_options-max_total_connections:

    .. data:: max_total_connections

        Specifies the maximum number of active connections.
        This option affects libcurl `CURLMOPT_MAX_TOTAL_CONNECTIONS <https://curl.haxx.se/libcurl/c/CURLMOPT_MAX_TOTAL_CONNECTIONS.html>`_.

        **Note**

        You may want to control the maximum number of sockets that a particular HTTP client uses simultaneously.
        If a system passes many requests to distinct hosts, then libcurl cannot reuse sockets.
        In this case, setting ``max_total_connections`` may be useful
        since it causes curl to avoid creating too many sockets, which would not be used anyway.

..  _client_object:

client_object
~~~~~~~~~~~~~

..  class:: client_object

    An HTTP client instance that exposes the API for :ref:`making requests <making_requests>`.
    To create the client, call :ref:`http.client.new() <http-new>`.

    ..  _client_object-request:

    ..  method:: request(method, url, body, opts)

        Make an HTTP request and receive a response.

>         :param string method: a request HTTP method, for example, ``GET``, ``POST``, or ``PUT``

Reviewer 3: I think we need all possible values in the reference, not just examples.

Author: agree

    ..  _request_options-low_speed_time:

    .. data:: low_speed_time

        The time that the transfer speed should be below the "low speed limit" for the library to consider it too slow and abort.

        :rtype: integer

        **See also:** `CURLOPT_LOW_SPEED_TIME <https://curl.haxx.se/libcurl/c/CURLOPT_LOW_SPEED_TIME.html>`_

    ..  _request_options-max_header_name_len:

    .. data:: max_header_name_len

        The maximum length of a header name.
        If a header name length exceeds this value, it is truncated to this length.
        The default value is ``32``.

        :rtype: integer

    ..  _request_options-follow_location:

    .. data:: follow_location

        Specify whether the HTTP client follows redirect URLs provided in the ``Location`` header for ``3xx`` responses.
        When a non-``3xx`` response is received, the client returns it as a result.
        If you set this option to ``false``, the client returns the first ``3xx`` response.

        :rtype: boolean

        **See also:** :ref:`Redirects <response_redirect>`

    ..  _request_options-no_proxy:

    .. data:: no_proxy

>         A comma-separated list of hosts that do not require proxies, or '*', or ''.

Reviewer 3: ```suggestion
        A comma-separated list of hosts that do not require proxies, or ``'*'``, or ``''``.
```

Just an idea: the code formatting can improve readability here

Reviewer 3: Same in the next option

    .. data:: headers

        Response headers.

        :rtype: table

        **See also:** :ref:`Headers <response_headers>`

    ..  _response_object-cookies:

    .. data:: cookies

        Response cookies.
        The value is an array of two elements where the first one is the
        cookie value and the second one is an array with the cookie's options.

        :rtype: table

        **See also:** :ref:`Cookies <response_cookies>`

    ..  _response_object-body:

    .. data:: body

        A response body.
        Use :ref:`decode <response_object-decode>` to decode the response body.

        :rtype: table

        **See also:** :ref:`Response body <response_body>`

    ..  _response_object-proto:

    .. data:: proto

        An HTTP protocol version.

        :rtype: string

    ..  _response_object-decode:

    .. function:: decode()

        Decode the response body to a Lua object based on the content type.

        :return: a decoded body
        :rtype: table

        **See also:** :ref:`Deserialization <response_deserialization>`

        **Since:** :doc:`2.11.0 </release/2.11.0>`

.. _io_object:

io_object
~~~~~~~~~

..  class:: io_object

>     An IO object used to read/write data in chunks.

Reviewer 3: Google [doesn't like](https://developers.google.com/style/slashes#slashes-with-alternatives) such use of the slash. Me neither %)
